rfcs/resource-prototypes.md

```diff
@@ -0,0 +1,237 @@
+
+## Posts [/posts]
+
+### List posts [GET]
+
++ Response 200
+    + Attributes (array[Post], required, fixed-type)
+
++ Response 429 (application/json)
+    + Attributes
+        + status: tooManyRequests (string, required, fixed)
+        + waitFor: `10` (number, required) - wait before next request
+
++ Response 400 (application/json)
+    + Attributes
+        + status: badRequest (string, required, fixed)
+        + errors (array[string], required, fixed-type) - list of errors in request
+
+### Show post [GET]
+
++ Response 200
+    + Attributes (Post, required)
+
++ Response 429 (application/json)
+    + Attributes
+        + status: tooManyRequests (string, required, fixed)
+        + waitFor: `10` (number, required) - wait before next request
+
++ Response 400 (application/json)
+    + Attributes
+        + status: badRequest (string, required, fixed)
+        + errors (array[string], required, fixed-type) - list of errors in request
+
+## Users [/users]
+
+### List users [GET]
+
++ Response 200
+    + Attributes (array[User], required, fixed-type)
+
++ Response 429 (application/json)
+    + Attributes
+        + status: tooManyRequests (string, required, fixed)
+        + waitFor: `10` (number, required) - wait before next request
+
++ Response 400 (application/json)
+    + Attributes
+        + status: badRequest (string, required, fixed)
+        + errors (array[string], required, fixed-type) - list of errors in request
+
++ Response 403 (application/json)
+    + Attributes
+        + status: forbidden (string, required, fixed)
+
+### Create user [POST]
+
++ Response 200
+    + Attributes (User, required)
+
++ Response 429 (application/json)
+    + Attributes
+        + status: tooManyRequests (string, required, fixed)
+        + waitFor: `10` (number, required) - wait before next request
+
++ Response 400 (application/json)
+    + Attributes
+        + status: badRequest (string, required, fixed)
+        + errors (array[string], required, fixed-type) - list of errors in request
+
++ Response 403 (application/json)
+    + Attributes
+        + status: forbidden (string, required, fixed)
+```
+
+As you can see responses 429, 400 and 403 are duplicated several times. It is
+possible to extract response description into
+[Data Structures](https://github.com/apiaryio/api-blueprint/blob/master/API%20Blueprint%20Specification.md#def-data-structures)
+section and it will help to reduce duplication but you have to write something
+like
+
+```apib
++ Response 429 (application/json)
+    + Attributes (TooManyRequestsError)
+```
+
+for every resource. It is possible to remove duplication completely using new
+proposed syntax:
+
+```apib
+# Data Structures
+
+## Post
+
++ title: `test` (string, required)
++ body: `hello world` (string, required)
+
+## User
+
++ name: `John Smith` (string, required)
++ email: `user@example.com` (string, required)
+
+## Authorized (Resource Prototype)
+
++ Response 403 (application/json)
```

Comment:
@kylef imho you are trying to solve another problem. I see two independent problems:

1. do not repeat identical responses (problem which I want to solve)
2. decouple mson from http protocol (problem which you want to solve)

I think we shouldn't mix these problems. I just want to solve small problem which would help to reduce api description dramatically. Decoupling is a bigger problem which can help to describe different API (not only JSON HTTP API) and it requires much more changes so I think it's a good idea to solve it now. My changes do not affect decoupling at all because it just repeats current response structure in new section. When we find solution for bigger problem it will be easy to make changes in resource section and the same changes in resource prototype section but for now it would be logically to make the same structure in resource and resource prototype sections.

I thinks `Data Structures` is not the best place for `Resource Prototype` items so I prefer to create another section called `Resource Prototypes` and move all resource prototypes there. It will be more logically for users and also easier to implement.

@kylef @pksunkara what do you think about it?